File: random_forest/go/rf.go
package main

import (
    "fmt"
    "math/rand"
    "os"
    "strings"
)

type DataPoint struct {
    Label int
    Features []float64
}

func LoadData(filename string) []DataPoint {
    data := []DataPoint{}
    file, err := os.Open(filename)
    if err != nil {
        fmt.Println(err)
        return data
    }
    defer file.Close()

    scanner := bufio.NewScanner(file)
    for scanner.Scan() {
        line := scanner.Text()
        features := strings.Split(line, ",")
        label := int(features[0])
        features = features[1:]
        features = make([]float64, len(features))
        for i, f := range features {
            features[i] = float64(f)
        }

        data = append(data, DataPoint{Label: label, Features: features})
    }

    return data
}

func CreateRandomForest(numTrees int) []int {
    treeLabels := make([]int, numTrees)
    for i := 0; i < numTrees; i++ {
        treeLabels[i] = rand.Intn(2)
    }

    return treeLabels
}

func ClassifyData(data []DataPoint, treeLabels []int) []int {
    predictions := make([]int, len(data))
    for i, dp := range data {
        correctLabel := dp.Label
        prediction := 0
        for _, treeLabel := range treeLabels {
            if treeLabel == dp.Label {
                prediction++
            }
        }

        predictions[i] = prediction > numTrees/2 ? 1 : 0
    }

    return predictions
}

func CalculateAccuracy(predictions []int, data []DataPoint) float64 {
    correct := 0
    for i, prediction := range predictions {
        if prediction == data[i].Label {
            correct++
        }
    }

    accuracy := float64(correct) / float64(len(data))
    return accuracy * 100
}

func main() {
    data := LoadData("data.csv")
    treeLabels := CreateRandomForest(10)
    predictions := ClassifyData(data, treeLabels)
    accuracy := CalculateAccuracy(predictions, data)
    fmt.Printf("Accuracy: %.2f%%\n", accuracy)
}

File: random_forest/c/rf.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

#define NUM_TREES 10
#define NUM_FEATURES 5

typedef struct {
  int label;
  float features[NUM_FEATURES];
} DataPoint;

DataPoint *data = NULL;
int num_data_points = 0;

void load_data() {
  FILE *fp = fopen("data.csv", "r");
  if (fp == NULL) {
    printf("Error opening data file.\n");
    exit(1);
  }

  num_data_points = 0;
  while (1) {
    DataPoint *dp = malloc(sizeof(DataPoint));
    int i;
    for (i = 0; i < NUM_FEATURES; i++) {
      fscanf(fp, "%f,", &dp->features[i]);
    }
    fscanf(fp, "%d\n", &dp->label);

    num_data_points++;
    if (feof(fp)) {
      break;
    }

    data = realloc(data, sizeof(DataPoint) * num_data_points);
    data[num_data_points - 1] = *dp;
  }

  fclose(fp);
}

void free_data() {
  if (data != NULL) {
    free(data);
    data = NULL;
  }
}

int main() {
  load_data();

  // Create a random forest with 10 trees.
  int *tree_labels = malloc(sizeof(int) * NUM_TREES);
  for (int i = 0; i < NUM_TREES; i++) {
    tree_labels[i] = rand() % 2;
  }

  // Classify each data point using the random forest.
  int *predictions = malloc(sizeof(int) * num_data_points);
  for (int i = 0; i < num_data_points; i++) {
    int correct_label = data[i].label;
    int prediction = 0;
    for (int j = 0; j < NUM_TREES; j++) {
      if (tree_labels[j] == data[i].label) {
        prediction++;
      }
    }

    predictions[i] = prediction > NUM_TREES / 2 ? 1 : 0;
  }

  // Calculate the accuracy of the random forest.
  int correct = 0;
  for (int i = 0; i < num_data_points; i++) {
    if (predictions[i] == data[i].label) {
      correct++;
    }
  }

  float accuracy = (float)correct / num_data_points;
  printf("Accuracy: %.2f%%\n", accuracy * 100);

  free_data();
  free(tree_labels);
  free(predictions);

  return 0;
}
